### ISSUE TITLE
Way to customize individual markers added to cluster manager?

### ISSUE THREAD

**Issue by s-p-a-r-k** [googlemaps/google-maps-ios-utils#21]
Is there a way to change icon and title of individual markers when they are not clustered? For Android, I know I can extend DefaultClusterRenderer and override onBeforeClusterItemRendered method, but I cannot find a way to do this for iOS

**Comment by mountainvat** [comment]
What you can do is extend GMUDefaultClusterRenderer and override this method 
```
- (GMSMarker *)markerWithPosition:(CLLocationCoordinate2D)position
                             from:(CLLocationCoordinate2D)from
                         userData:(id)userData
                      clusterIcon:(UIImage *)clusterIcon
                         animated:(BOOL)animated
```

This is where the marker is actually created. I've included some sample code here. Please note the new code is between the "// new code" markers. Let us know how you go. We will plan to make a better API for this use case in a later release.
```
  CLLocationCoordinate2D initialPosition = animated ? from : position;
  GMSMarker *marker = [GMSMarker markerWithPosition:initialPosition];
  marker.userData = userData;
  if (clusterIcon != nil) {
    marker.icon = clusterIcon;
    marker.groundAnchor = CGPointMake(0.5, 0.5);
  }
 // new code starts here.
  else {
   // this is new code for customizing individual markers
    marker.icon =  [self getCustomIconForItem:userData];  // stub method
    marker.title = [self getCustomTitleItem:userData];  // stub method
  }
// new code ends here.

  marker.map = _mapView;

  if (animated) {
    [CATransaction begin];
    [CATransaction setAnimationDuration:kGMUAnimationDuration];
    marker.layer.latitude = position.latitude;
    marker.layer.longitude = position.longitude;
    [CATransaction commit];
  }
```

**Comment by s-p-a-r-k** [comment]
@mountainvat This works for changing all of individual markers to the same icon and title, but I do not know how to use this code to change only some of individual markers. I would like to set the icon and the title of markers depending on a variable stored in my custom GMUClusterItem. Is there a way to do this too?

**Comment by mountainvat** [comment]
one way you can do replace this code:

```
    marker.icon =  [self getCustomIconForItem:userData];  // stub method
    marker.title = [self getCustomTitleItem:userData];  // stub method

```
with
```
    marker.icon =  [self getCustomIconForItem:userData] ? :  marker.icon;  // stub method
    marker.title = [self getCustomTitleItem:userData] ? : marker.title;  // stub method

```
and in the getCustomXXX method, you'd return nil if the item should not be customized.

Hope it helps.

**Comment by s-p-a-r-k** [comment]
@mountainvat I am sorry, but I'm not familiar with Objective C and don't know how I should replace `[self getCustomIconForItem:userData]` and `[self getCustomTitleItem:userData]`.
If I want to title my marker with POIItem's name (from [https://developers.google.com/maps/documentation/ios-sdk/utility/marker-clustering](https://developers.google.com/maps/documentation/ios-sdk/utility/marker-clustering)), for example, should I call userData's name variable?
`        
          
    /// Point of Interest Item which implements the GMUClusterItem protocol.
    class POIItem: NSObject, GMUClusterItem {
        var position: CLLocationCoordinate2D
        var name: String!
          
        init(position: CLLocationCoordinate2D, name: String) {
            self.position = position
            self.name = name
          }
        }`

**Comment by mountainvat** [comment]
You can safely cast userData to your cluster item object. So I'd imagine something like this:

```
- (NSString *)getCustomTitleItem:(id)userData {
  POIItem *item = userData;
  return item.name;
}
```

**Comment by s-p-a-r-k** [comment]
@mountainvat I tried to put your code in GMSDefaultClusterRenderer.m, but it says use of undeclared identifier. Should I move POIItem class to Clustering folder that has all Google-Maps-IOS-Utils classes?

**Comment by mountainvat** [comment]
I would not put the code inside GMSDefaultClusterRenderer.m. I'd rather extend it in your own application's classes. Since the custom renderer lives in your app's target it should have access yo your GMUClusterItem classes. So you could do something like this:

```
@interface MyClusterRenderer : GMSDefaultClusterRenderer
@end

```
```
@implementation MyClusterRenderer

- (NSString *)getCustomTitleItem:(id)userData {
  YourOwnGMUClusterItemClass *item = userData;
  return item.name;
}

@end
```

**Comment by s-p-a-r-k** [comment]
@mountainvat That makes sense. Apparently, I did not read your very first answer that said to extend GMSDefaultClusterRenderer. However, it fails to build. ("Command failed due to signal: Segmentation fault: 11"
This is what I have for my custom renderer:

    #import <Foundation/Foundation.h>
    #import "CustomRenderer.h"
    #import "MyProjectName-swift.h"

    // Animation duration for marker splitting/merging effects.
    static const double kGMUAnimationDuration = 0.5;  // seconds.

    @implementation CustomRenderer
    // Map view to render clusters on.
    __weak GMSMapView *_mapView;

    - (NSString *)getCustomTitleItem:(id)userData {
        MyOwnClusterItemClass *item = userData;
        return item.title;
    }

    - (GMSMarker *)markerWithPosition:(CLLocationCoordinate2D)position
                                 from:(CLLocationCoordinate2D)from
                             userData:(id)userData
                          clusterIcon:(UIImage *)clusterIcon
                             animated:(BOOL)animated {
        CLLocationCoordinate2D initialPosition = animated ? from : position;
        GMSMarker *marker = [GMSMarker markerWithPosition:initialPosition];
        marker.userData = userData;
        if (clusterIcon != nil) {
            marker.icon = clusterIcon;
            marker.groundAnchor = CGPointMake(0.5, 0.5);
        } else {
            marker.title = [self getCustomTitleItem:userData] ? : marker.title;
        }
        marker.map = _mapView;
    
        if (animated) {
            [CATransaction begin];
            [CATransaction setAnimationDuration:kGMUAnimationDuration];
            marker.layer.latitude = position.latitude;
            marker.layer.longitude = position.longitude;
            [CATransaction commit];
        }
        return marker;
    }

    @end 


Do you know what is causing this error?

**Comment by mountainvat** [comment]
I cannot reason why it would cause a Segmentation fault. Do you have a stacktrace?

Anyway I can see a problem with your code (maybe not related): the _mapView instance var is always nil. What you also need to do is override initWithMapView to capture the mapView into your private _mapView variable.

Add this method to your new derived class:
```

- (instancetype)initWithMapView:(GMSMapView *)mapView
           clusterIconGenerator:(id<GMUClusterIconGenerator>)iconGenerator {
  self = [super initWithMapView:mapView clusterIconGenerator:iconGenerator];
  if (self) {
    _mapView = mapView;
  }
}
```

**Comment by mountainvat** [comment]
@s-p-a-r-k : Any luck with this? Can we close the ticket?

**Comment by s-p-a-r-k** [comment]
@mountainvat Sorry, I haven't had a chance to try it. But you can close it. I will tell you if it works. Thank you very much for your help!

**Comment by vjerryv** [comment]
@mountainvat I've tried to override GMUDefaultClusterRenderer, but I got this error.
 
Duplicate definition of category 'GoogleMaps' on interface 'GMSCoordinateBounds'

![screenshot 2016-09-22 11 40 37](https://cloud.githubusercontent.com/assets/3232807/18736432/7c44a02c-80b9-11e6-8f0c-548b35985023.png)

**Comment by astrolope** [comment]
I must be numb-skulled, I'm trying to over-ride GMUDefaultClusterRenderer to achieve more dynamicity in my custom markers without having to edit the main file. I copied the code from the original file to a new class as previously aforementioned in the previous post. Although I'm now receiving mach-o linker errors after setting it up.  Deleting my custom class completely removes the errors.  I've managed to change the marker fine, but am struggling on how to do it "the right way".

**Comment by mountainvat** [comment]
FYI: we are working on making marker customization easier. Will have a release out end of this week or earlier next week. So stay tuned.

**Comment by unlimitedK** [comment]
I have gone through the new  GMUClusterRendererDelegate 

How can I use this in swift, I have to show markerimage(which would again be changing to different conditions) and have a label at the bottom of this marker. In Swift, I have used func renderer(renderer: GMUClusterRenderer, markerForObject object: AnyObject) -> GMSMarker

but it is not getting called before placing the marker so that I can change images on that?

**Comment by mountainvat** [comment]
If you update to version 1.1.x there should be new events for before and after the markers are added to map. Please see this post for more information: https://github.com/googlemaps/google-maps-ios-utils/blob/master/CustomMarkers.md

**Comment by kwstasna** [comment]
@mountainvat could you provide a swift example for this GMUDefaultClusterRenderer delegate?

**Comment by jamesynotnot** [comment]
Here's how I do it with a custom cluster item:

enum MarkerType: Int32 {

    case RegularMarker, LandmarkMarker, GovernmentMarker

}


// Clustering markers customization

class CustomClusterItem: NSObject, GMUClusterItem {

    var position: CLLocationCoordinate2D

    var type: MarkerType

    var icon: UIImage



    init(position: CLLocationCoordinate2D, type: MarkerType, icon: UIImage)
{

        self.position = position

        self.type = type

        self.icon = icon

    }

}

    //MARK: - GMUClusterRendererDelegate

    func renderer(_ renderer: GMUClusterRenderer, willRenderMarker marker:
GMSMarker) {

        if (marker.userData! is CustomClusterItem) {

            let customClusterItem = (marker.userData! as! CustomClusterItem)



            switch customClusterItem.type {

                case MarkerType.RegularMarker:

                    marker.icon = appDelegate.markerImage

                    break

                case MarkerType.LandmarkMarker:

                    marker.icon = appDelegate.landmarkImage

                    break

                case MarkerType.GovernmentMarker:

                    marker.icon = appDelegate.govtImage

                    break

            }

        }

    }

**Comment by kwstasna** [comment]
@jamesynotnot actually wanted to translate somehow this file

https://github.com/googlemaps/google-maps-ios-utils/blob/master/app/CustomMarkerViewController.m

Which has the functions to make this view happen.
Because im in the same position to have a json with names and images urls.

**Comment by jamesynotnot** [comment]
https://objectivec2swift.com/

**Comment by kwstasna** [comment]
@jamesynotnot not the best way 😞

**Comment by kayschmitt** [comment]
@jamesynotnot I cant get that render function to work. Could you post your whole controller by chance?

**Comment by jamesynotnot** [comment]
The controller is quite large and customized, so can't post it.  Your
controller should be like this:

class ViewController: UIViewController, GMUClusterManagerDelegate,
GMUClusterRendererDelegate, GMSMapViewDelegate {

}


Let me know what's failing and I can try to help

**Comment by hansemannn** [comment]
Hey guys, just came across this during my PR to support clustering with google maps in Titanium ([here](https://github.com/hansemannn/ti.googlemaps/pull/58)). A convenient way would be to just subclass it and use the protocol to create everything. Unfortunately, methods like `markerWithPosition:from:userData:clusterIcon:animated:` are not in the public interface, so they won't be suggested. That would be one of the improvements. Rather then that, things links handling the mapView reference should be solved using a reference to the super class, so we don't need to declare own initializers. Not sure if I can help here, but this sounds like a doable change! :-)

**Comment by BeetLab** [comment]
@mountainvat - But what about that :

/**
 * Raised when a marker (for a cluster or an item) is about to be added to the map.
 * Use the marker.userData property to check whether it is a cluster marker or an
 * item marker.
 */
- (void)renderer:(id<GMUClusterRenderer>)renderer willRenderMarker:(GMSMarker *)marker;

in @protocol GMUClusterRendererDelegate<NSObject>?

**Comment by mountainvat** [comment]
@BeetLab you can use the event to customize your mark before it's added to the map. See https://github.com/googlemaps/google-maps-ios-utils/blob/master/CustomMarkers.md for an example of how to use it.

**Comment by prsreejith** [comment]
How to load this cluster GMSMapview into custom view like below ? 

- (void)loadView {
  GMSCameraPosition *camera =
      [GMSCameraPosition cameraWithLatitude:kCameraLatitude longitude:kCameraLongitude zoom:10];
  _mapView = [GMSMapView mapWithFrame:CGRectZero camera:camera];
  self.myMapView = _mapView;
}

currently its loading only to self.view

**Comment by grundmanise** [comment]
@mountainvat Hello,

in this function:

`func renderer(_ renderer: GMUClusterRenderer, willRenderMarker marker: GMSMarker) {}`

`makrer.iconView` is always `nil` even if I set it up before like this:

  ```
  func renderer(_ renderer: GMUClusterRenderer, willRenderMarker marker: GMSMarker) {
        
        // Check if marker or cluster
        if marker.userData is PlaceMarker {

            if let userData = marker.userData as? PlaceMarker {
               marker.iconView = MarkerView(caption: userData.caption)
            }

            marker.groundAnchor = CGPoint(x: 0.5, y: 1)
            marker.isFlat = true
            marker.appearAnimation = kGMSMarkerAnimationPop

        } else {
            // Apply custom view for cluster
            marker.iconView = ClusterViewIcon()
            
            // Show clusters above markers
            marker.zIndex = 1000;
            marker.groundAnchor = CGPoint(x: 0.5, y: 1)
            marker.isFlat = true
            marker.appearAnimation = kGMSMarkerAnimationPop
        }
    }
```

how can one implement a `guard` to only setup `iconView` and other `marker` properties only when the marker is rendered for the first time? otherwise it is just a waste of resources..

**Comment by mountainvat** [comment]
How many different cluster views do you have? If you have only 1 custom view (shared by many clusters) then I'd suggest return the same instance in ClusterViewIcon() so that the view is only rendered once.

In addition, this event is only raised once when the marker is brought into view unless a recluster happens (say on a zoom level change). For this maybe raise a separate issue and we will see if we can address it in the coming release.

**Comment by grundmanise** [comment]
I did it before - stored a pre-created view in a `var`,  and assigned this `var` to the `iconView`, thus the view was rendered only once. But now I need to setup captions. Each marker has it's own caption. So I can't use 1 view for every marker.

Exactly. On the zoom level change. If nothing changes (i.e. no clustering / declustering happens), we recreate already rendered markers and waste resources.

**Comment by miragessee** [comment]
@jamesynotnot 

If don't call func renderer:

implementation GMUClusterRendererDelegate

renderer.delegate = self

**Comment by Prathapreddy26** [comment]
@mountainvat Can you help with my problem? I am done with the clustering. The markers are clustered. but, it is not showing the Cluster count. Any help will be appreciable.
